### PR TITLE
[CHIA-1313] Port NFT RPCs to `@marshal` decorator

### PR DESCRIPTION
**Pull request by Quexington**
Another PR following https://github.com/Chia-Network/chia-blockchain/pull/18651

Three things to note in this PR:
1. A comical amount of code in the NFT wallet commands was untested.  The large number of changes to that section of the code have to do with a) the confusing dictionary structure it was using (replaced in favor of lists of the RPC types) and b) actually fixing a legitimate issue that was not caught due to the lack of coverage: https://github.com/Chia-Network/chia-blockchain/pull/19617/files#diff-3980198ee87efa35f7bb4b7f1184457cba0a4df3eb430222cf96d476d91aa364L1472-R1500
2. Much of the parsing logic in the NFT RPCs was flexible with allowing things like `hash`, `metadata_hash` or `license_hash` to be not 32-bytes.  However, in compliance with the CHIP defining these structures:
```
Data Hash: The SHA-256 hash of the NFT’s content. For example, if the NFT is a .jpeg file, then the Data Hash is the hash of that file. Upon the NFT's minting, this hash is recommended to match the hash of the content of the first item of the Data URI list (defined above). This hash shall not be modified after the NFT's minting
Metadata Hash: A SHA-256 hash of the contents of the first item of the Metadata URI list. This hash shall not be modified after the NFT's minting
License Hash: A SHA-256 hash of the contents of the first item of the License URI list. The Transfer Program shall dictate whether this hash may be modified after the NFT's minting
```
the hashes have been solidified to only allow 32-byte values.  Interestingly, our tests were relying on this not being the case, but I don't believe that's a reason to support it going forward.
3. The parsing logic in the RPCs also frequently had logic that seemed to assume that a given entry in the request dictionary could be bytes.  This would allow someone to directly specify the puzzle hash rather than an address.  However, it's impossible to send bytes directly over JSON so this was unreachable code and was deleted as part of simplification.

### CODE REVIEW

File: chia/_tests/cmds/cmd_test_utils.py [hunk 1]
```diff
@@ -183,17 +188,21 @@ async def get_nft_info(self, coin_id: str, latest: bool = True) -> dict[str, Any
...
-        return NFTWallet.royalty_calculation(
-            royalty_assets_dict=royalty_assets_dict,
-            fungible_asset_dict=fungible_asset_dict,
+        request: NFTCalculateRoyalties,
+    ) -> NFTCalculateRoyaltiesResponse:
+        self.add_to_log("nft_calculate_royalties", (request,))
+        return NFTCalculateRoyaltiesResponse.from_json_dict(
+            NFTWallet.royalty_calculation(
+                {
+                    asset.asset: (asset.royalty_address, uint16(asset.royalty_percentage))
```
Comment (altendky) [inline chia/_tests/cmds/cmd_test_utils.py]: it seems to already be a `uint16`.

```suggestion
                    asset.asset: (asset.royalty_address, asset.royalty_percentage)
```

File: chia/rpc/wallet_request_types.py [hunk 2]
```diff
@@ -367,6 +408,103 @@ class NFTGetWalletsWithDIDsResponse(Streamable):
...
+class NFTGetInfoResponse(Streamable):
+    nft_info: NFTInfo
+
+
+# utility for NFTCalculateRoyalties
+@streamable
+@dataclass(frozen=True)
+class RoyaltyAsset(Streamable):
+    asset: str
+    royalty_address: str
```
Comment (altendky) [inline chia/rpc/wallet_request_types.py]: strings not bytes?

File: chia/_tests/cmds/wallet/test_nft.py [hunk 3]
```diff
@@ -197,17 +192,25 @@ def test_nft_add_uri(capsys: object, get_test_cli_clients: tuple[TestRpcClients,
...
             self,
-            wallet_id: int,
-            nft_coin_id: str,
-            key: str,
-            uri: str,
-            fee: int,
+            request: NFTAddURI,
             tx_config: TXConfig,
-            push: bool,
+            extra_conditions: tuple[Condition, ...] = tuple(),
```
Comment (altendky) [inline chia/_tests/cmds/wallet/test_nft.py:199]: adding this here without adding the value to the log seems to risk someone submitting some `extra_conditions` and forgetting to verify them.  leaving the parameter missing as before, if i understood this correctly, means they would get an error when passing a value for this which avoids the risk.  i think it should perhaps remain missing or get added to the logging.

whatever we decide here would presumably apply to several similar cases in this pr.

File: chia/_tests/cmds/wallet/test_offer.toffer [hunk 4]
```diff
@@ -1 +1 @@
...
...k0teysu0k4tnphlthdc28dl9gur3zsd2gd8f08jxptjxmdg3aylltk6m4a4ywj63e5ttmf4ky6e2xqq8mgwl0mjsfq36g3x2kxprj5pcshnawu3kqjmjnkjl0lgxgktxvt6u2q7tv7825hdae8gymcufdkac76dnjxv7z9j3wkz3ffyf8uctsgwfmp877hn0wy435j8a5rz6gfc86n77vety2unz4maggz6u8kkewcf4qxan47jx62jmgthtgj0hjp773u5ywk7g58ek34xh4emr69eerdl6jjpuz0ke6v34mlpa6p4qzwq0gxjvpe3jp5uq7sdyqgxn5r77yljzl2xns87g8vnjlu9g5tdfalnswfxfwy4h8nf60jmwz4lw6yalmqrwd2ezgml74qpdyf3cmt7ffgv3qekejgjef0lmwe8mrgkxgks28e83w57chz7dd89se56hu2r7wmrz233sfd43km92gtjrwawqrc2s7s37r83736cnt60kuz4m5cxsevjp98xdq5gmhcuje8s7ygduqy7m5ew72n4kt6vm4mxfrj3k6w7lgpp23et92ynpapvynlulnpywp8r8hxnegf9900kpyryvtnknj6nluusrw4k375amvmxejrvfvpss4zmzyan5yvs5m88a49gp9ruuvl0869cmxxlxxryhnuj73qjhzevnatf2uy4pk3kvsk04efhdex67dx04gj6q66q66q67q2uq2acwq2a0mzhgr8m0pglfudn7c2pgzpg288tc9vgfc7k9ppql6vcc0q6wvvn003gss5k36l2fvfspv4dv8gn2e42qc7qgfpkurmpxrtd3vm6gcjg7n4aapavjjyyezqszdnmzuyc3e4zgd0pe4zrafw96xs3a6a2q06q86q86q86q86q8cv6q86q8lmu4h33lgsru0npefsxa24qpxf9yet2mckt7u8t93unjhk433txw5zuf87h45n8jdq5fpa5fthc5qgeqhk2zsfcsperh25s6qqn29hy97nsmykdvwp52geg7meme6uj6c59mjx5r50s45kdt04hufr5zy9pw5s7zklpzge48msufj6s7rd5m89elrjm20uyx2fcud4xr8nuky8xx7fuxhcu2p942x5t7lf6nlvvkc42hfz47jv9mn0prhe9a4ule3sfhlqxezpumzsm3ayd8
```
Comment (altendky) [inline chia/_tests/cmds/wallet/test_offer.toffer:1]: i'm going to ask, very generically and very lazily...  `:]`  why did this change?

File: chia/cmds/wallet_funcs.py [hunk 5]
```diff
@@ -1209,24 +1231,26 @@ async def mint_nft(
...
-                CMDTXConfigLoader(
+                request=NFTMintNFTRequest(
+                    wallet_id=uint32(wallet_id),
+                    royalty_address=royalty_address,
+                    target_address=target_address,
+                    hash=bytes32.from_hexstr(hash),
+                    uris=uris,
+                    meta_hash=bytes32.from_hexstr(metadata_hash) if metadata_hash is not None else None,
+                    meta_uris=metadata_uris,
+                    license_hash=bytes32.from_hexstr(license_hash) if license_hash is not None else None,
```
Comment (altendky) [inline chia/cmds/wallet_funcs.py:1242]: this may be ok, but technically this is a cli breaking change by not handling empty string `""` hashes as well as `None`.

File: chia/cmds/wallet_funcs.py [hunk 6]
```diff
@@ -538,19 +552,21 @@ async def make_offer(
                     print()
                     print(f"Including Fees: {Decimal(fee) / units['chia']} XCH, {fee} mojos")
 
-                if royalty_asset_dict != {}:
-                    royalty_summary: dict[Any, list[dict[str, Any]]] = await wallet_client.nft_calculate_royalties(
-                        royalty_asset_dict, fungible_asset_dict
+                if royalty_assets != []:
```
Comment (altendky) [inline chia/cmds/wallet_funcs.py]: since we're changing it anyways, maybe?

```suggestion
                if len(royalty_assets) > 0:
```

File: chia/cmds/wallet_funcs.py [hunk 7]
```diff
@@ -1469,11 +1497,7 @@ def driver_dict_asset_is_nft_supporting_royalties(driver_dict: dict[bytes32, Puz
 
 def driver_dict_asset_is_fungible(driver_dict: dict[bytes32, PuzzleInfo], asset_id: bytes32) -> bool:
     asset_dict: PuzzleInfo = driver_dict[asset_id]
-    return not asset_dict.check_type(
-        [
-            AssetType.SINGLETON.value,
-        ]
-    )
+    return not asset_dict.type() == AssetType.SINGLETON.value
```
Comment (altendky) [inline chia/cmds/wallet_funcs.py]: :raised_eyebrow: 

```suggestion
    return asset_dict.type() != AssetType.SINGLETON.value
```

File: chia/wallet/trade_manager.py [hunk 8]
```diff
@@ -862,6 +862,16 @@ async def respond_to_offer(
             complete_offer, valid_spend_solver = await self.check_for_final_modifications(
                 Offer.aggregate([offer, take_offer]), solver, inner_action_scope
             )
+
+        async with action_scope.use() as interface:
+            if interface.side_effects.get_unused_derivation_record_result is not None:
+                # This error is of a protection against potential misues of this band-aid solution.
+                # We should put more thought into how sub-action scopes are generated and what effects
+                # we might want to push. A ticket to this respect can be found [CHIA-2984].
```
Comment (altendky) [inline chia/wallet/trade_manager.py]:
```suggestion
                # This error is a protection against potential misues of this band-aid solution.
                # We should put more thought into how sub-action scopes are generated and what effects
                # we might want to push. A ticket to this respect can be found at [CHIA-2984].
```

File: chia/wallet/trade_manager.py [hunk 9]
```diff
@@ -862,6 +862,16 @@ async def respond_to_offer(
             complete_offer, valid_spend_solver = await self.check_for_final_modifications(
                 Offer.aggregate([offer, take_offer]), solver, inner_action_scope
             )
+
+        async with action_scope.use() as interface:
```
Comment (altendky) [inline chia/wallet/trade_manager.py]: would this be appropriate in a standalone pr?

File: chia/wallet/wallet_action_scope.py [hunk 10]
```diff
@@ -165,5 +165,5 @@ async def new_wallet_action_scope(
         extra_spends=self.side_effects.extra_spends,
         singleton_records=self.side_effects.singleton_records,
     )
-    if self.side_effects.get_unused_derivation_record_result is not None:
+    if push and self.side_effects.get_unused_derivation_record_result is not None:
```
Comment (altendky) [inline chia/wallet/wallet_action_scope.py]: would this be appropriate in a standalone pr?

File: chia/wallet/wallet_state_manager.py [hunk 11]
```diff
@@ -369,6 +369,16 @@ def get_wallet(self, id: uint32, required_type: type[TWalletType]) -> TWalletTyp
 
         return wallet
 
+    @asynccontextmanager
+    async def puzzle_hash_db_writer(self) -> AsyncIterator[None]:
```
Comment (altendky) [inline chia/wallet/wallet_state_manager.py]: would this be appropriate in a standalone pr?

File: chia/_tests/wallet/test_wallet.py [hunk 12]
```diff
@@ -2133,6 +2145,24 @@ async def test_forced_new_puzzle_hash(self, wallet_environments: WalletTestFrame
...
         assert len(list(set(coin.puzzle_hash for coin in tx.spend_bundle.additions()))) == 2
 
+    @pytest.mark.parametrize(
+        "wallet_environments",
+        [{"num_environments": 1, "blocks_needed": [1], "reuse_puzhash": True, "trusted": True}],
+        indirect=True,
+    )
+    @pytest.mark.limit_consensus_modes
+    @pytest.mark.anyio
+    async def test_puzzle_hashes_not_committed(self, wallet_environments: WalletTestFramework) -> None:
```
Comment (altendky) [inline chia/_tests/wallet/test_wallet.py]: would this be appropriate in a standalone pr?

File: chia/_tests/cmds/wallet/test_wallet.py [hunk 13]
```diff
@@ -246,9 +251,9 @@ async def get_wallet_balance(self, wallet_id: int) -> dict[str, uint64]:
                 "unconfirmed_wallet_balance": uint64(0),
             }
 
-        async def get_nft_wallet_did(self, wallet_id: uint8) -> dict[str, Optional[str]]:
-            self.add_to_log("get_nft_wallet_did", (wallet_id,))
-            return {"did_id": "0xcee228b8638c67cb66a55085be99fa3b457ae5b56915896f581990f600b2c652"}
+        async def get_nft_wallet_did(self, request: NFTGetWalletDID) -> NFTGetWalletDIDResponse:
+            self.add_to_log("get_nft_wallet_did", (request.wallet_id,))
+            return NFTGetWalletDIDResponse("did:chia:1qgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpq4msw0c")
```
Comment (altendky) [inline chia/_tests/cmds/wallet/test_wallet.py:256]: is this indicating we are changing the interface from hex to bech32?

File: chia/rpc/wallet_rpc_client.py [hunk 14]
```diff
@@ -986,211 +1001,97 @@ async def create_new_nft_wallet(self, did_id: Optional[str], name: Optional[str]
...
+        request: NFTMintNFTRequest,
         tx_config: TXConfig,
-        meta_hash: Optional[str] = "",
-        meta_uris: list[str] = [],
-        license_hash: Optional[str] = "",
-        license_uris: list[str] = [],
-        edition_total: Optional[int] = 1,
-        edition_number: Optional[int] = 1,
-        fee: int = 0,
-        royalty_percentage: int = 0,
```
Comment (altendky) [inline chia/rpc/wallet_rpc_client.py]: looks like we're changing from percentage to amount?

File: chia/_tests/wallet/rpc/test_wallet_rpc.py [hunk 15]
```diff
@@ -1665,24 +1679,69 @@ async def have_nfts() -> bool:
...
+                        None,
+                        uint64(10000),
+                    ),
+                    FungibleAsset(
+                        None,
+                        uint64(11111),
+                    ),
+                ],
+            )
+        )
```
Comment (altendky) [inline chia/_tests/wallet/rpc/test_wallet_rpc.py:1726]: would this be appropriate in a standalone pr?

File: chia/rpc/wallet_rpc_api.py [hunk 16]
```diff
@@ -2993,160 +3022,136 @@ async def did_transfer_did(
...
+        nft_wallet = self.service.wallet_state_manager.get_wallet(id=request.wallet_id, required_type=NFTWallet)
+        if request.royalty_amount == 10000:
             raise ValueError("Royalty percentage cannot be 100%")
-        if isinstance(royalty_address, str):
-            royalty_puzhash = decode_puzzle_hash(royalty_address)
-        elif royalty_address is None:
+        if request.royalty_address is not None:
+            royalty_puzhash = decode_puzzle_hash(request.royalty_address)
+        else:
             royalty_puzhash = await action_scope.get_puzzle_hash(self.service.wallet_state_manager)
```
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py]: i think we're losing the previously supported 'puzhash is the address' case.

File: chia/rpc/wallet_rpc_api.py [hunk 17]
```diff
@@ -2993,160 +3022,136 @@ async def did_transfer_did(
...
+            ("lu", request.license_uris),
+            ("sn", request.edition_number),
+            ("st", request.edition_total),
         ]
-        if "meta_hash" in request and len(request["meta_hash"]) > 0:
-            metadata_list.append(("mh", hexstr_to_bytes(request["meta_hash"])))
-        if "license_hash" in request and len(request["license_hash"]) > 0:
-            metadata_list.append(("lh", hexstr_to_bytes(request["license_hash"])))
+        if request.meta_hash is not None:
+            metadata_list.append(("mh", request.meta_hash))
```
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py]: existing code was requiring the meta hash to be greater than zero to run this.  it can't be zero anymore so i get that we don't check that.  but...  could someone have done silliness with this now impossible case of a zero-length meta hash?  as in, is this a relevant interface change to be considered?

File: chia/rpc/wallet_rpc_api.py [hunk 18]
```diff
@@ -2993,160 +3022,136 @@ async def did_transfer_did(
...
+        if request.license_hash is not None:
+            metadata_list.append(("lh", request.license_hash))
```
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py]: as above

File: chia/rpc/wallet_rpc_api.py [hunk 19]
```diff
@@ -2993,160 +3022,136 @@ async def did_transfer_did(
...
-            "success": True,
-            "spend_bundle": None,  # tx_endpoint wrapper will take care of this
-            "nft_id": nft_id_bech32,
-            "transactions": None,  # tx_endpoint wrapper will take care of this
-        }
+        return NFTMintNFTResponse(
+            [],
+            [],
+            wallet_id=request.wallet_id,
+            spend_bundle=WalletSpendBundle([], G2Element()),  # tx_endpoint wrapper will take care of this
```
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py]: should we have a `WalletSpendBundle.empty` or somesuch maybe?  or just make `None` ok.  or...  but, feel free to close this once you've read it.  i think the pattern pre-exists sufficiently here to be considered separately.

File: chia/rpc/wallet_rpc_api.py [hunk 20]
```diff
@@ -3457,161 +3427,116 @@ async def nft_get_info(self, request: dict[str, Any]) -> EndpointResult:
...
-        key = request["key"]
-        nft_coin_id = request["nft_coin_id"]
-        if nft_coin_id.startswith(AddressType.NFT.hrp(self.service.config)):
-            nft_coin_id = decode_puzzle_hash(nft_coin_id)
+        nft_wallet = self.service.wallet_state_manager.get_wallet(id=request.wallet_id, required_type=NFTWallet)
+        if request.nft_coin_id.startswith(AddressType.NFT.hrp(self.service.config)):
+            nft_coin_id = decode_puzzle_hash(request.nft_coin_id)
         else:
-            nft_coin_id = bytes32.from_hexstr(nft_coin_id)
+            nft_coin_id = bytes32.from_hexstr(request.nft_coin_id)
```
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py]: someday, a helper for this pattern might be worthwhile.  `:]`  feel free to close this after reading.

File: chia/rpc/wallet_rpc_api.py [hunk 21]
```diff
@@ -3457,161 +3427,116 @@ async def nft_get_info(self, request: dict[str, Any]) -> EndpointResult:
...
+                ("st", meta.edition_total),
             ]
-            if "meta_hash" in meta and len(meta["meta_hash"]) > 0:
-                nft_metadata.append(("mh", hexstr_to_bytes(meta["meta_hash"])))
-            if "license_hash" in meta and len(meta["license_hash"]) > 0:
-                nft_metadata.append(("lh", hexstr_to_bytes(meta["license_hash"])))
+            if meta.meta_hash is not None:
+                nft_metadata.append(("mh", meta.meta_hash))
+            if meta.license_hash is not None:
+                nft_metadata.append(("lh", meta.license_hash))
```
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py]: a couple more cases where the interface is no longer accepting an empty string.  is that a case we need to support?

File: chia/rpc/wallet_rpc_api.py [hunk 22]
```diff
@@ -3457,161 +3427,116 @@ async def nft_get_info(self, request: dict[str, Any]) -> EndpointResult:
...
-        xch_coins = None
-        if xch_coin_list:
-            xch_coins = {Coin.from_json_dict(xch_coin) for xch_coin in xch_coin_list}
-        xch_change_target = request.get("xch_change_target", None)
-        if xch_change_target is not None:
-            if xch_change_target[:2] == "xch":
-                xch_change_ph = decode_puzzle_hash(xch_change_target)
+        target_list = [decode_puzzle_hash(target) for target in request.target_list]
+        if request.xch_change_target is not None:
+            if request.xch_change_target[:2] == "xch":
```
Comment (altendky) [inline chia/rpc/wallet_rpc_api.py]: i think we have a (pre-existing) bug here.  this can never be true.  (also why having a parser and checking the hrp or whatever is better than string processing `:]`)

```python-console
Python 3.12.7 (main, Nov 20 2024, 20:26:30) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> "xch"[:2]
'xc'
```

File: chia/_tests/wallet/nft_wallet/test_nft_wallet.py [hunk 23]
```diff
@@ -98,7 +107,10 @@ async def test_nft_wallet_creation_automatically(wallet_environments: WalletTest
         wallet_node_0.wallet_state_manager, wallet_0, name="NFT WALLET 1"
     )
     metadata = Program.to(
-        [("u", ["https://www.chia.net/img/branding/chia-logo.svg"]), ("h", "0xD4584AD463139FA8C0D9F68F4B59F185")]
+        [
+            ("u", ["https://www.chia.net/img/branding/chia-logo.svg"]),
+            ("h", "0xD4584AD463139FA8C0D9F68F4B59F185D4584AD463139FA8C0D9F68F4B59F185"),
```
Comment (altendky) [inline chia/_tests/wallet/nft_wallet/test_nft_wallet.py:112]: is there a case where the shorter value is appropriate to use?  i'm guessing this changed because the new interface is requiring a 'full address'.

File: chia/_tests/wallet/nft_wallet/test_nft_wallet.py [hunk 24]
```diff
@@ -512,6 +527,19 @@ async def test_nft_wallet_creation_and_transfer(wallet_environments: WalletTestF
     await time_out_assert(30, get_nft_count, 1, nft_wallet_0)
     await time_out_assert(30, get_nft_count, 1, nft_wallet_1)
 
+    # Test an error case
+    with pytest.raises(ResponseFailureError, match="The NFT doesn't support setting a DID."):
```
Comment (altendky) [inline chia/_tests/wallet/nft_wallet/test_nft_wallet.py:531]: would this be appropriate in a standalone pr?

File: chia/_tests/wallet/nft_wallet/test_nft_wallet.py [hunk 25]
```diff
@@ -2045,26 +2193,42 @@ async def test_nft_bulk_transfer(wallet_environments: WalletTestFramework) -> No
...
 
     async with wallet_1.wallet_state_manager.new_action_scope(wallet_environments.tx_config, push=True) as action_scope:
         wallet_1_ph = await action_scope.get_puzzle_hash(wallet_1.wallet_state_manager)
     fee = uint64(1000)
     address = encode_puzzle_hash(wallet_1_ph, AddressType.XCH.hrp(env_1.node.config))
+    with pytest.raises(ResponseFailureError, match="You can only transfer"):
+        await env_0.rpc_client.transfer_nft_bulk(
+            NFTTransferBulk(target_address=address, nft_coin_list=[nft_coin_list[0]] * (MAX_NFT_CHUNK_SIZE + 1)),
+            wallet_environments.tx_config,
+        )
```
Comment (altendky) [inline chia/_tests/wallet/nft_wallet/test_nft_wallet.py:2231]: would this be appropriate in a standalone pr?

**Review by altendky — CHANGES_REQUESTED** [review]

File: chia/_tests/cmds/cmd_test_utils.py (hunk 1 above)
Comment (Quexington) [reply chia/_tests/cmds/cmd_test_utils.py]: fixed

File: chia/_tests/cmds/wallet/test_nft.py (hunk 3 above)
Comment (Quexington) [reply chia/_tests/cmds/wallet/test_nft.py:199]: Yeah I would prefer these match the client exactly so I'll add it to the logs.

File: chia/_tests/cmds/wallet/test_offer.toffer (hunk 4 above)
Comment (Quexington) [reply chia/_tests/cmds/wallet/test_offer.toffer:1]: This is part of increasing test coverage.  It's a more complex offer that exercises more edge cases.  It's extracted from the `test_complex_nft_offer` test.

File: chia/_tests/cmds/wallet/test_wallet.py (hunk 13 above)
Comment (Quexington) [reply chia/_tests/cmds/wallet/test_wallet.py:256]: No, this test framework (ever a pain in my behind) was just returning an incorrectly formatted mock value.

File: chia/_tests/wallet/nft_wallet/test_nft_wallet.py (hunk 23 above)
Comment (Quexington) [reply chia/_tests/wallet/nft_wallet/test_nft_wallet.py:112]: Like I said in the description, I'm leaning on the CHIP definition here as justification for breaking the backwards compatibility.  Bringing us into compliance.  So my direct answer to your question is, no.

File: chia/_tests/wallet/nft_wallet/test_nft_wallet.py (hunk 24 above)
Comment (Quexington) [reply chia/_tests/wallet/nft_wallet/test_nft_wallet.py:531]: It certainly could but this came as the result of reviewing test coverage once the PR was complete.  I'd have to actually do more work to convert this test to the old format before PRing it and then re-do my changes on merge.  I figured methodologically it made sense to include it in this PR.

File: chia/_tests/wallet/nft_wallet/test_nft_wallet.py (hunk 25 above)
Comment (Quexington) [reply chia/_tests/wallet/nft_wallet/test_nft_wallet.py:2231]: It certainly could but this came as the result of reviewing test coverage once the PR was complete. I'd have to actually do more work to convert this test to the old format before PRing it and then re-do my changes on merge. I figured methodologically it made sense to include it in this PR.

File: chia/_tests/wallet/rpc/test_wallet_rpc.py (hunk 15 above)
Comment (Quexington) [reply chia/_tests/wallet/rpc/test_wallet_rpc.py:1726]: It certainly could but this came as the result of reviewing test coverage once the PR was complete. I'd have to actually do more work to convert this test to the old format before PRing it and then re-do my changes on merge. I figured methodologically it made sense to include it in this PR.

File: chia/_tests/wallet/test_wallet.py (hunk 12 above)
Comment (Quexington) [reply chia/_tests/wallet/test_wallet.py]: hmm there's a git issue here, because this does in fact exist in main already.

File: chia/cmds/wallet_funcs.py (hunk 6 above)
Comment (Quexington) [reply chia/cmds/wallet_funcs.py]: sure, fixed

File: chia/cmds/wallet_funcs.py (hunk 5 above)
Comment (Quexington) [reply chia/cmds/wallet_funcs.py:1242]: Yeah I kind of figured it would be fine.

File: chia/cmds/wallet_funcs.py (hunk 7 above)
Comment (Quexington) [reply chia/cmds/wallet_funcs.py]: hmm, fixed

File: chia/rpc/wallet_request_types.py (hunk 2 above)
Comment (Quexington) [reply chia/rpc/wallet_request_types.py]: Yeah it's bech32

File: chia/rpc/wallet_rpc_api.py (hunk 16 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: Yeah, there's a couple instances like that in this PR, I guess I should explain.  The original tree made no sense.  The first branch checked if it was a string, the second checked if it was `None` and then the third branch assumed it was a bytes32.  Except you can't send bytes over JSON... so... not sure why it was in there, but it was indeed uncovered by tests and impossible to reach.

File: chia/rpc/wallet_rpc_api.py (hunk 17 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: Yeah this is rolled into the call out in the description that these values are now forced to be 32 bytes according to the relevant CHIP.

File: chia/rpc/wallet_rpc_api.py (hunk 18 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: comment as above as well :)

File: chia/rpc/wallet_rpc_api.py (hunk 19 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: I would not mind such a property, but I do not think `None` should be allowed because in the final request it certainly will not be.

File: chia/rpc/wallet_rpc_api.py (hunk 20 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: Agreed, I have thought while doing these ports that a `PuzzlehashOrAddress` type might be warranted in a lot of places.  I haven't looked into how painful that would be.  (Now YOU resolve it)

File: chia/rpc/wallet_rpc_api.py (hunk 21 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: Per the description, these values are being hardened to be _only_ 32 bytes (or `None`)

File: chia/rpc/wallet_rpc_api.py (hunk 22 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: fixed

File: chia/rpc/wallet_rpc_client.py (hunk 14 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_client.py]: Yeah lol, we're not, but the RPC poorly named this value and that's the value we have to protect for backwards compatibility.  Once it gets to `NFTWallet` it is once again called `_percentage`.

File: chia/wallet/trade_manager.py (hunk 9 above)
Comment (Quexington) [reply chia/wallet/trade_manager.py]: Yeah, git error sorry, it will disappear next push.

File: chia/wallet/trade_manager.py (hunk 8 above)
Comment (Quexington) [reply chia/wallet/trade_manager.py]: I'm not going to do this because this will also disappear from the diff 😅

File: chia/wallet/wallet_action_scope.py (hunk 10 above)
Comment (Quexington) [reply chia/wallet/wallet_action_scope.py]: Yeah sorry, this has been extracted and this is github's fault for showing you (and mine for not merging from main sooner (I actually did this _because_ you were in the middle of reviewing lol))

File: chia/wallet/wallet_state_manager.py (hunk 11 above)
Comment (Quexington) [reply chia/wallet/wallet_state_manager.py]: See the comments about how this will disappear 🪄

File: chia/cmds/wallet_funcs.py [hunk 26]
```diff
@@ -759,17 +775,20 @@ async def take_offer(
...
+                    RoyaltyAsset(
+                        encode_puzzle_hash(royalty_asset_id, AddressType.NFT.hrp(config)),
+                        encode_puzzle_hash(address, AddressType.XCH.hrp(config)),
+                        percentage,
+                    )
                 )
 
-        if royalty_asset_dict != {}:
-            fungible_asset_dict: dict[Any, uint64] = {}
+        if royalty_assets != []:
```
Comment (altendky) [inline chia/cmds/wallet_funcs.py]:
```suggestion
        if len(royalty_assets) > 0:
```

File: chia/cmds/wallet_funcs.py [hunk 27]
```diff
@@ -780,20 +799,24 @@ async def take_offer(
...
                             nft_royalty_currency = result[1]
-                    fungible_asset_dict[nft_royalty_currency] = uint64(requested[fungible_asset_id_str])
+                    fungible_assets.append(
+                        FungibleAsset(nft_royalty_currency, uint64(requested[fungible_asset_id_str]))
+                    )
 
-            if fungible_asset_dict != {}:
-                royalty_summary: dict[Any, list[dict[str, Any]]] = await wallet_client.nft_calculate_royalties(
-                    royalty_asset_dict, fungible_asset_dict
+            if fungible_assets != []:
```
Comment (altendky) [inline chia/cmds/wallet_funcs.py]:
```suggestion
            if len(fungible_assets) > 0:
```

File: chia/rpc/wallet_rpc_api.py (hunk 22 above)
Comment (altendky) [reply chia/rpc/wallet_rpc_api.py]: extra note that i didn't think about before...  this can't be leveraged on testnet what with not using the hrp logic.  `:]`  again, welcome to be handled separately.

**Review by altendky — CHANGES_REQUESTED** [review]

File: chia/rpc/wallet_rpc_api.py (hunk 22 above)
Comment (Quexington) [reply chia/rpc/wallet_rpc_api.py]: ticket made: https://chia-network.atlassian.net/browse/CHIA-3206

**Review by altendky — APPROVED** [review]

**Review by markelrod — APPROVED** [review]
